src/parser.js

```diff
@@ -40,554 +37,391 @@ export class ParserImplementation {
         } else {
-          values.push(this.parseAccessOrCallScope(name, token));
+          result = new CallFunction(result, args);
         }
+        context = 0;
         break;
-      case T$StringLiteral:
-      case T$NumericLiteral:
-        keys.push(name);
+      case T$TemplateTail:
+        result = new LiteralTemplate([this.tokenValue], [], [this.tokenRaw], result);
         this.nextToken();
-        this.expect(T$Colon);
-        values.push(this.parseExpression());
         break;
-      default:
-        this.error(`Unexpected token ${this.tokenRaw}`);
-      }
-      if (this.currentToken !== T$RBrace) {
-        this.expect(T$Comma);
+      case T$TemplateContinuation:
+        result = this.parseTemplate(context | C$Tagged, result);
+      // no default
       }
     }
 
-    this.expect(T$RBrace);
-
-    return new LiteralObject(keys, values);
+    return result;
   }
 
-  parseExpressionList(terminator) {
-    let result = [];
+  parseTemplate(context, func) {
+    const cooked = [this.tokenValue];
+    const raw = context & C$Tagged ? [this.tokenRaw] : undefined;
+    this.expect(T$TemplateContinuation);
+    const expressions = [this.parseExpression()];
 
-    if (this.currentToken !== terminator) {
-      do {
-        result.push(this.parseExpression());
-      } while (this.optional(T$Comma));
+    while ((this.currentToken = this.scanTemplateTail()) !== T$TemplateTail) {
+      cooked.push(this.tokenValue);
+      if (context & C$Tagged) {
+        raw.push(this.tokenRaw);
+      }
+      this.expect(T$TemplateContinuation);
+      expressions.push(this.parseExpression());
     }
 
-    return result;
+    cooked.push(this.tokenValue);
+    if (context & C$Tagged) {
+      raw.push(this.tokenRaw);
+    }
+    this.nextToken();
+    return new LiteralTemplate(cooked, expressions, raw, func);
   }
 
   nextToken() {
-    return this.currentToken = this.scanToken();
-  }
-
-  nextChar() {
-    return this.currentChar = this.input.charCodeAt(++this.index);
-  }
-
-  scanToken() {
-    while (this.hasNext) {
-      // skip whitespace.
-      if (this.currentChar <= $SPACE) {
+    /*
+     * Each index in CharScanners (0-65535) contains a scan function for the charCode with that number.
+     * The array is "zero-filled" with a throwing function and has functions for all ASCII chars (except ~@#`\)
+     * and IdentifierParts from the Latin1 script (1314 in total).
+     * Additional characters can be added via addIdentifierStart / addIdentifierPart.
+     */
+    while (this.index < this.length) {
+      if (this.currentChar <= /*whitespace*/0x20) {
         this.nextChar();
         continue;
       }
-
-      this.lastIndex = this.startIndex;
       this.startIndex = this.index;
-
-      // handle identifiers and numbers.
-      if (isIdentifierStart(this.currentChar)) {
-        return this.scanIdentifier();
+      if (this.currentChar === /*$*/0x24 || (this.currentChar >= /*a*/0x61 && this.currentChar <= /*z*/0x7A)) {
```

Comment:
Does this mean `_` is not considered an identifier ?

Reply:
It is. As explained in the comment right below that, that `if` statement (and the one further up for whitespace) is purely a performance tweak for the most common characters. Comparisons are several times faster than an array indexer. The logic itself is redundant.

You could comment them out and reduce the `while` loop in `nextToken()` code to just this, and it would still work:
```js
while (this.index < this.length) {
  this.startIndex = this.index;
  if ((this.currentToken = CharScanners[this.currentChar](this)) !== null) { // a null token means the character must be skipped
    return;
  }
}
```